aten/src/ATen/native/cuda/MultinomialKernel.cu

```diff
@@ -228,72 +234,80 @@ sampleMultinomialOnce(int64_t* dest,
     }
 
     int chunks = (categories + (int)blockDim.x - 1) / blockDim.x;
-    scalar_t prevHighProb = zero;
+    accscalar_t prevHighProb = accZero;
     found = false;
 
     for (int chunk = 0; chunk < chunks && !found; ++chunk) {
       // All threads in bounds load a value
       int cat = chunk * blockDim.x + threadIdx.x;
 
-      accscalar_t a_dist_val = cat < categories ?
-                               static_cast<accscalar_t>(dist[curDist * stride_dist + cat * stride_categories]) / sum :
-                               accZero;
-      scalar_t dist_val = static_cast<scalar_t>(a_dist_val);
+      accscalar_t dist_val = cat < categories ?
+                             static_cast<accscalar_t>(dist[curDist * stride_dist + cat * stride_categories]) / sum :
+                             accZero;
 
-      smem[threadIdx.x] = dist_val;
+      if (is_valid_thread) {
```

Comment:
why are changes with `is_valid_thread` and `valid_block_size` necessary here? Previous implementation was able to handle invalid threads (when number of categories was larger than maximum block size, the last chunk only had a few valid threads). 
There's very little harm in doing the number of iterations later based of actual block size and not valid block size, as kernel should still be bandwidth bound, and iterations with smem accesses are fast.

Reply:
If I remember it correctly, I did it so as not to increase the amount of shared memory. Thinking back, I guess it doesn't make much sense, since we'll only use `sizeof(dtype)` per thread. So, I will increase the amount of shared memory, and remove both `is_valid_thread` and `valid_block_size`.

Reply:
Shared memory requirements of this kernel are pretty small and don't limit occupancy in any way.